src/link.cpp

```diff
@@ -1653,15 +1653,25 @@ static void construct_linker_job_elf(LinkJob *lj) {
         soname = buf_sprintf("lib%s.so.%" ZIG_PRI_usize, buf_ptr(g->root_out_name), g->version_major);
     }
 
+    if (target_requires_pie(g->zig_target) && !is_dyn_lib && g->libc != nullptr) {
+        lj->args.append("-pie");
+    }
+
     lj->args.append("-o");
     lj->args.append(buf_ptr(&g->output_file_path));
 
     if (lj->link_in_crt) {
         const char *crt1o;
         if (g->zig_target->os == OsNetBSD) {
             crt1o = "crt0.o";
+        } else if (target_is_android(g->zig_target)) {
+            crt1o = "crtbegin_dynamic.o";
```

Comment:
Instead of the logic below, shouldn't it be here?

```c++
if (g->have_dynamic_link) {
    crt1o = "crtbegin_dynamic.o";
} else {
    crt1o = "crtbegin_static.o";
}
```

Then the other logic remains unchanged:

```c++
         } else if (!g->have_dynamic_link) { 
             crt1o = "crt1.o";
         }
```